src/new_api_v04/predicate.rs

```diff
@@ -19,80 +19,62 @@
 //! for <var2> is in the returned set.
 
 use crate::HashSet;
-use std::hash::Hash;
+
+/// A predicate with a fixed arity N.
+pub trait ArityPredicate {
+    /// Get Predicate arity
+    fn arity(&self) -> usize;
+}
 
 /// A predicate for pattern matching.
-#[derive(Clone, Debug)]
+///
+/// Assign predicates must be of arity N >= 1 and bind the variable passed as its
+/// first argument.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
```

Comment:
I'd add a comment about `Assign < Filter`, since it's a strong requirement for the constraint ordering.